### ISSUE TITLE
Type errors could be improved

### ISSUE THREAD

**Issue by Namek** [mint-lang/mint#187]
## Visually

![image](https://user-images.githubusercontent.com/31058/74608461-89eb8a00-50e1-11ea-99fb-fe85fcb3f73a.png)

## Informational

I would really see a **difference** between expected type and given type. The way it is given is time consuming. Having type system this strong makes it very possible to present type diffs.

The example above is:

### input

```
state puzzle : Puzzle = {
  islands =
    {
      list = [],
      fields = Map.empty()
    },
  connections =
    {
      list = [],
      fields = Map.empty()
    },
  width = 0,
  height = 0,
  maxConnectionCount = 2
}
```

### found type

```
(
  islands: Islands(list: Array(Island(index: Number, maxConnectionCounts: ConnectionSizes(top: Number, right: Number, bottom: Number, left: Number), currentConnectionCounts: ConnectionSizes(top: Number, right: Number, bottom: Number, left: Number))), fields: Map(Int, Bool)),
  connections: Islands(list: Array(Island(index: Number, maxConnectionCounts: ConnectionSizes(top: Number, right: Number, bottom: Number, left: Number), currentConnectionCounts: ConnectionSizes(top: Number, right: Number, bottom: Number, left: Number))), fields: Map(Int, Bool)),
  width: Number,
  height: Number,
  maxConnectionCount: Number)
```

### expected type: `Puzzle`

```
record Puzzle {
  islands : Islands,
  connections : Connections,
  width : Number,
  height : Number,
  maxConnectionCount : Number
}
```
with other records:

```
record Island {
  index : Number,
  maxConnectionCounts : ConnectionSizes,
  currentConnectionCounts : ConnectionSizes
}

record Islands {
  list : Array(Island),
  fields : Map(Int, Bool)
}

record Connection {
  idx1 : Number,
  idx2 : Number,
  connectionSize : Number,
  orientation : Orientation
}

record Connections {
  list : Array(Connection),
  fields : Map(Int, IndexPair)
}

record IndexPair {
  idx1 : Int,
  idx2 : Int
}

record ConnectionSizes {
  top : Number,
  right : Number,
  bottom : Number,
  left : Number
}
```

**Comment by gdotdesign** [comment]
Yes I agree this can be better, unfortunately this is something that needs to be done in a case-by-case bases.

A little background: when matching records we try to match them to a record definition and this error is the result of not finding one, since it's in a state its type checker can catch this error and produce an other more meaningful one.

**Comment by gdotdesign** [comment]
Resolved with 06e08eda3ea909fb015acd5de1a6ade02c7ecd56
